File: trunk/quote/Projects/PluginModelIO/TemplateSaver.vb
Imports System.Data.SqlClient
Imports System.Transactions

Imports Model.Template

Imports DB
Imports DB.QuoteDataBaseTableAdapters
Imports System.Windows.Forms

Public Class TemplateSaver

    Public Function Save(ByVal q As Header) _
                        As Integer

        System.Windows.Forms.Cursor.Current = Cursors.WaitCursor

        ' Ensure the properies are updated
        Host.App.MainForm.Focus()

        Dim o As PrimaryProperties = q.PrimaryProperties

        Dim newId As Integer
        Dim id As Integer = o.CommonId

        Dim adaptor As New QuoteDataBaseTableAdapters._QuoteTableAdapter
        If id > 0 Then
            adaptor.Connection.Open()
            o.CommonLastModified = Date.Now
            Dim test As Integer = adaptor.Update( _
                o.Customer.Name, o.RequestForQuoteNumber, _
                o.PartNumber, False, Nothing, _
                o.CommonInitials, o.CommonCreatedDate, _
                o.LastModified, o.Customer.ID, _
                o.CommonId, o.CommonId)
            adaptor.Connection.Close()
            newId = id
        Else
            adaptor.Connection.Open()
            Dim proxy As New QuoteTableProxy(adaptor)
            proxy.Transaction = adaptor.Connection.BeginTransaction
            adaptor.Insert(o.Customer.Name, _
                o.RequestForQuoteNumber, o.PartNumber, False, 0, _
                o.CommonInitials, Date.Now, Date.Now, o.Customer.ID)
            Dim cmd As SqlCommand = New SqlCommand( _
                "SELECT @@IDENTITY", adaptor.Connection)
            cmd.Transaction = proxy.Transaction
            newId = CInt(cmd.ExecuteScalar())
            proxy.Transaction.Commit()
            adaptor.Connection.Close()
            If id = 0 Then
                q.PrimaryProperties.SetID(newId)
            End If
        End If

        adaptor.Connection.Open()
        CommonSaver.DeleteProperties(newId)
        CommonSaver.SaveNoteProperties(newId, q.NoteProperties)
        CommonSaver.SaveOtherProperties(newId, q.OtherProperties, True)
        CommonSaver.SaveComputationProperties(newId, q.ComputationProperties, True)
        CommonSaver.DeleteComponents(newId)
        CommonSaver.SaveComponents(q, newId, True)
        adaptor.Connection.Close()

        q.PrimaryProperties.SendEvents()

        q.ComputationProperties.ClearDirty()
        q.OtherProperties.ClearDirty()
        q.PrimaryProperties.ClearDirty()
        q.ClearDirty()

        System.Windows.Forms.Cursor.Current = Cursors.Default

        Return newId
    End Function

End Class

File: trunk/quote/Projects/PluginModelIO/TemplateSaver.cs
namespace Model.IO
{
    using System;
    using System.Data.SqlClient;
    using System.Globalization;
    using System.Linq;
    using System.Windows.Forms;
    
    using DB;
    
    using Model.Template;

    public static class TemplateSaver
    {
        public static int Save(Common.Header header)
        {
            System.Windows.Forms.Cursor.Current = Cursors.WaitCursor;

            // Ensure the properies are updated
            Host.App.MainForm.Focus();

            Model.Template.PrimaryProperties o = 
                header.PrimaryProperties as Model.Template.PrimaryProperties;

            int newId = 0;
            int id = o.CommonId;

            DB.QuoteDataBaseTableAdapters._QuoteTableAdapter adaptor = 
                new DB.QuoteDataBaseTableAdapters._QuoteTableAdapter();
            
            if (id > 0) 
            {
                adaptor.Connection.Open();
                o.CommonLastModified = System.DateTime.Now;
                
                adaptor.Update(
                    o.Customer.Name, 
                    o.RequestForQuoteNumber, 
                    o.PartNumber, 
                    false, 
                    null, 
                    o.CommonInitials, 
                    o.CommonCreatedDate, 
                    o.LastModified, 
                    o.Customer.Id, 
                    o.CommonId,
                    o.CommonId);
                
                adaptor.Connection.Close();
                newId = id;
            } 
            else 
            {
                adaptor.Connection.Open();
                QuoteTableProxy proxy = new QuoteTableProxy(adaptor);
                proxy.Transaction = adaptor.Connection.BeginTransaction();
                
                adaptor.Insert(
                    o.Customer.Name, 
                    o.RequestForQuoteNumber, 
                    o.PartNumber, 
                    false, 
                    0, 
                    o.CommonInitials, 
                    System.DateTime.Now, 
                    System.DateTime.Now, 
                    o.Customer.Id);
                
                SqlCommand cmd = new SqlCommand("SELECT @@IDENTITY", adaptor.Connection);
                cmd.Transaction = proxy.Transaction;
                newId = Convert.ToInt32(cmd.ExecuteScalar(), CultureInfo.CurrentCulture);
                proxy.Transaction.Commit();
                adaptor.Connection.Close();
                
                if (id == 0) 
                {
                    header.PrimaryProperties.SetId(newId);
                }
            }

            adaptor.Connection.Open();
            CommonSaver.DeleteProperties(newId);
            CommonSaver.SaveNoteProperties(newId, header.NoteProperties);
            CommonSaver.SaveOtherProperties(newId, header.OtherProperties, true);
            CommonSaver.SaveComputationProperties(newId, header.ComputationProperties, true);
            CommonSaver.DeleteComponents(newId);
            CommonSaver.SaveComponents(header, newId, true);
            adaptor.Connection.Close();

            header.ComputationProperties.IsDirty = false;
            header.OtherProperties.IsDirty = false;
            header.PrimaryProperties.IsDirty = false;
            header.IsDirty = false;

            System.Windows.Forms.Cursor.Current = Cursors.Default;

            return newId;
        }
    }
}
